### ISSUE TITLE
Unit tray.target not found

### ISSUE THREAD

**Issue by jlesquembre** [nix-community/home-manager#2064]
### Issue description

On current master, I get the error `Unit tray.target not found` when I try to start some systemd units (like udiskie or blueman).
I'm using waybar, I think this error is related to #2027 

#### Maintainer CC
@berbiche @NickHu

**Comment by berbiche** [comment]
Unfortunately, the tray.target is only created if xsession management is enabled in HM.

The creation of said systemd target should be moved elsewhere.

In the meantime, you can create the systemd target manually from the xsession.nix? file and add `Install.WantedBy = [ "graphical-session-pre.target" ]` to the target definition.

**Comment by sumnerevans** [comment]
> Unfortunately, the tray.target is only created if xsession management is enabled in HM.

Ah, this explains why I encountered that error on my wayland system. We should definitely fix that.

**Comment by zeorin** [comment]
I have this problem on X11. `flameshot` includes a service that wants the `tray.target`.

**Comment by vyorkin** [comment]
Can confirm the problem on X11 with `flameshot`, subscribed to this issue exactly for this reason

**Comment by zeorin** [comment]
@vyorkin I should probably mention that I worked around this for the time being with the following in my `home.nix`:
```nix
	systemd.user.targets.tray = {
		Unit = {
			Description = "Home Manager System Tray";
			Requires = [ "graphical-session-pre.target" ];
		};
	};
```
which I lifted from `home-manager`'s source somewhere.

**Comment by DavHau** [comment]
Still an issue. The home-manager udiskie service doesn't work on my nixos 21.05.
I'm using i3wm on x11 and there is no tray target

**Comment by aru-hackZ** [comment]
What I noticed is that I can't start any service in `$XDG_CONFIG_HOME/systemd/user`, so services like the ones created from `services.emacs` or `services.gpg-agent` can't be started. Searching a bit I didn't find any issue directly talking about it. `systemd-analyze --user unit-paths` shows `$XDG_CONFIG_HOME/systemd/user`, which makes the fact that it can't find services there even more strange.

**Comment by AshleyYakeley** [comment]
I believe this should be reopened? I'm using marco (rather than sway) without xsession management, and I'm coming across this issue.

**Comment by ncfavier** [comment]
I can see two solutions:
- change `Requires` directives on `tray.target` to `Wants` instead; @NickHu is there a specific reason you used `Requires`?
- define the `tray` target in the systemd module, regardless of whether xsession or sway is enabled

**Comment by Mikilio** [comment]
Just some thought to add:

I've been trying to wrap my head around what this target is for and how it's supposed to be used for. By the description it seems like the target being active indicates that there is a system tray active. This caused a lot of confusion for me because all 3rd party bars providing a system tray out there actually start after graphical-session.target creating dependency cycles on most desktops. Then again apps that use a system tray will request the tray from the dbus, which most likely will start the bar for the tray which makes their systemd service only a backup in case dbus hasn't started them yet. As far as I understand the only requirement for tray.target to be met seems to be dbus.service and possibly (xdg-desktop-portal.service ?). Am I right so far?

**Comment by Nebucatnetzer** [comment]
Wasn't able to start network manager applet because of the missing `tray.target`.

**Comment by arduano** [comment]
Interestingly, using @zeorin's fix for `flameshot`, it doesn't appear in the tray even though it runs fine in the background. It appears in the tray when I run `systemctl --user restart flameshot`.

**Comment by Kreyren** [comment]
Doesn't seem to work on NixOS 24.05 with GNOME using wayland either.

Cross-Referencing: https://github.com/NiXium-org/NiXium/pull/115

**Comment by ciferkey** [comment]
Confirming I'm seeing this on unstable now with `udiskie`. System info:

```
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.10.0, NixOS, 24.11 (Vicuna), 24.11.20240719.1d9c2c9`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - nixpkgs: `/etc/nix/path/nixpkgs`
```

Any other details I can provide that would help debug this?

**Comment by worthalex** [comment]
i too am encountering this issue on unstable after an update

```
activating the configuration...
setting up /etc...
reloading user units for alex...
restarting sysinit-reactivation.target
the following new units were started: systemd-hostnamed.socket
warning: the following units failed: home-manager-alex.service

× home-manager-alex.service - Home Manager environment for alex
     Loaded: loaded (/etc/systemd/system/home-manager-alex.service; enabled; preset: enabled)
     Active: failed (Result: exit-code) since Wed 2024-07-31 10:19:55 NZST; 70ms ago
   Duration: 4min 4.193s
 Invocation: 877dc7faf55849d3981a0504b9655bdb
    Process: 136236 ExecStart=/nix/store/877sla6xf6kngfzaabmqm4r5al918h9f-hm-setup-env /nix/store/5rkb4h56ggpqddf0v3v8r6zm7s5jqgql-home-manager-generation (code=exited, status=1/FAILURE)
   Main PID: 136236 (code=exited, status=1/FAILURE)
         IP: 0B in, 0B out
   Mem peak: 35.9M
        CPU: 853ms

Jul 31 10:19:55 sage-ice-cream hm-activate-alex[136236]: Activating reloadSystemd
Jul 31 10:19:55 sage-ice-cream hm-activate-alex[136642]: Starting units: udiskie.service
Jul 31 10:19:55 sage-ice-cream hm-activate-alex[136642]: Error: Error switching
Jul 31 10:19:55 sage-ice-cream hm-activate-alex[136642]: Caused by:
Jul 31 10:19:55 sage-ice-cream hm-activate-alex[136642]:     0: Failed to perform post-reload tasks
Jul 31 10:19:55 sage-ice-cream hm-activate-alex[136642]:     1: Failed to start unit udiskie.service
Jul 31 10:19:55 sage-ice-cream hm-activate-alex[136642]:     2: org.freedesktop.systemd1.NoSuchUnit: Unit tray.target not found.
Jul 31 10:19:55 sage-ice-cream systemd[1]: home-manager-alex.service: Main process exited, code=exited, status=1/FAILURE
Jul 31 10:19:55 sage-ice-cream systemd[1]: home-manager-alex.service: Failed with result 'exit-code'.
Jul 31 10:19:55 sage-ice-cream systemd[1]: Failed to start Home Manager environment for alex.
warning: error(s) occurred while switching to the new configuration
```

**Comment by KucharczykL** [comment]
Also happens with syncthing:

```nix
services.syncthing = {
  enable = true;
  tray = {
    enable = true;
    package = pkgs.syncthingtray;
};
```

```
pro 11 09:30:52 nixos hm-activate-lukas[41520]: Starting units: syncthingtray.service
pro 11 09:30:52 nixos hm-activate-lukas[41520]: Error: Error switching
pro 11 09:30:52 nixos hm-activate-lukas[41520]: Caused by:
pro 11 09:30:52 nixos hm-activate-lukas[41520]:     0: Failed to perform post-reload tasks
pro 11 09:30:52 nixos hm-activate-lukas[41520]:     1: Failed to start unit syncthingtray.service
pro 11 09:30:52 nixos hm-activate-lukas[41520]:     2: org.freedesktop.systemd1.NoSuchUnit: Unit tray.target not found.
pro 11 09:30:52 nixos systemd[1]: home-manager-lukas.service: Main process exited, code=exited, status=1/FAILURE
pro 11 09:30:52 nixos systemd[1]: home-manager-lukas.service: Failed with result 'exit-code'.
pro 11 09:30:52 nixos systemd[1]: Failed to start Home Manager environment for lukas.
pro 11 09:30:52 nixos systemd[1]: home-manager-lukas.service: Consumed 815ms CPU time, 29.7M memory peak, 2.3M read from disk.
pro 11 09:30:52 nixos nixos[40751]: switching to system configuration /nix/store/aw6zlnwvcgnfgm37kazarjnx909s6fak-nixos-system-nixos-25.05.20241203.55d15ad failed (status 4)
```

**Comment by LUVROK** [comment]
This strange problem appeared unexpectedly and still exists today, systemd.user.targets.tray helps

**Comment by onemoresuza** [comment]
#6332 was merged, which should fix this issue.

**Comment by khaneliman** [comment]
Fixed by #6332.
